src/rpm/headers/types.rs

```diff
@@ -32,30 +32,35 @@ pub struct PackageFileEntry {
     pub(crate) content: Vec<u8>,
 }
 
-#[non_exhaustive]
 #[derive(Copy, Clone, PartialEq, Eq, Debug, Hash)]
-pub enum FileMode {
-    // It does not really matter if we use u16 or i16 since all we care about
-    // is the bit representation which is the same for both.
-    Dir { permissions: u16 },
-    Regular { permissions: u16 },
-    SymbolicLink { permissions: u16 },
-    // For "Invalid" we use a larger integer since it is possible to create an invalid
-    // FileMode by providing an overflowing integer.
-    Invalid { raw_mode: i32, reason: &'static str },
+pub enum FileType {
+    Dir,
+    Regular,
+    SymbolicLink,
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, Debug, Hash)]
+pub struct FileMode {
+    file_type: FileType,
+    permissions: u16,
 }
 
 // there are more file types but in the context of RPM, only regular and directory and symbolic file should be relevant.
 // See <https://man7.org/linux/man-pages/man7/inode.7.html> section "The file type and mode"
 const FILE_TYPE_BIT_MASK: u16 = 0o170000; // bit representation = "1111000000000000"
 const PERMISSIONS_BIT_MASK: u16 = 0o7777; // bit representation = "0000111111111111"
-pub const REGULAR_FILE_TYPE: u16 = 0o100000; //  bit representation = "1000000000000000"
-pub const DIR_FILE_TYPE: u16 = 0o040000; //      bit representation = "0100000000000000"
-pub const SYMBOLIC_LINK_FILE_TYPE: u16 = 0o120000; // bit representation = "1010000000000000"
 
-// @todo: <https://github.com/rpm-rs/rpm/issues/52>
-impl From<u16> for FileMode {
-    fn from(raw_mode: u16) -> Self {
+const REGULAR_FILE_TYPE: u16 = 0o100000; //  bit representation = "1000000000000000"
+const DIR_FILE_TYPE: u16 = 0o040000; //      bit representation = "0100000000000000"
+const SYMBOLIC_LINK_FILE_TYPE: u16 = 0o120000; // bit representation = "1010000000000000"
+
+// @todo: turn permissions into a defined bitset
+// @todo: resolve ambiguity between mode and symlink - .symlink() ought to set the mode
```

Comment:
Maybe instead of `FileOptions::new()` we should have `FileOptions::regular(dest)`, `FileOptions::dir(dest)`, and `FileOptions::symlink(dest, link)` just like `FileMode`, and drop `.mode()`  on `FileOptionsBuilder` in favor of `.permissions()`

That would resolve the ambiguity.

Reply:
That's more clear than before, and for this `fn try_from(raw_mode: i32)`  is is possible to change `i32` to `u16`? so we can no need the runtime value check.